Adding initial version of `ww-rnaseq` pipeline (#304)
getwilds/wilds-wdl-library: pipelines/ww-rnaseq/README.md

# ww-rnaseq Pipeline
[![Project Status: Experimental – Useable, some support, not open to feedback, unstable API.](https://getwilds.org/badges/badges/experimental.svg)](https://getwilds.org/badges/#experimental)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)

> A comprehensive WILDS WDL pipeline for RNA-seq analysis, covering the core workflow from raw FASTQs through differential expression with integrated quality control at every stage. Inspired by nf-core/rnaseq but focused on the essential happy path.

Reviewer: I like the term 'happy path' but I'm not sure where it comes from or how it's defined exactly? Might be a bulk rna-seq thing?

Author: I think it's more of a software engineering term meaning "just the basic functionality without any edge cases/weirdness". To your point, it does seem a bit jargon-y, switched it to "focused on the core workflow steps" in https://github.com/getwilds/wilds-wdl-library/pull/304/commits/ae7819b6cfbe33d2654fe0af002dcd557d45faad

## Overview

This pipeline extends the simpler `ww-star-deseq2` pipeline with upstream read QC, adapter trimming, and aggregated MultiQC reporting. It integrates 7 WILDS WDL modules to perform a production-ready RNA-seq analysis workflow:

1. **Pre-trim QC** — Assess raw read quality
2. **Adapter Trimming** — Remove adapters and low-quality bases
3. **Post-trim QC** — Verify trimming improved read quality
4. **Genome Alignment** — Two-pass STAR alignment
5. **Alignment QC** — Comprehensive post-alignment quality metrics
6. **Differential Expression** — Count aggregation and DESeq2 analysis
7. **Aggregated Reporting** — MultiQC summary of all QC metrics

## Pipeline Structure

> **Complexity Level: Advanced** (8 steps, 7 distinct modules)

Reviewer: ```suggestion
**Complexity Level: Advanced** (9 steps, 7 distinct modules)
```

This pipeline is part of the [WILDS WDL Library](https://github.com/getwilds/wilds-wdl-library) and demonstrates:

- **Module Integration**: Combining 7 modules for a complete end-to-end RNA-seq workflow
- **Data Flow**: Seamless passing of outputs from QC through alignment to differential expression

> - **Best Practices**: Pre/post-trim QC, adapter trimming, two-pass alignment, and aggregated reporting

Reviewer: ```suggestion
```
I wouldn't focus on what it's demonstrating since this isn't an educational or proof-of-concept pipeline.

| `deseq2_pca_plot` | PCA plot of samples | ww-deseq2 |
| `deseq2_volcano_plot` | Volcano plot of differential expression | ww-deseq2 |
| `deseq2_heatmap` | Heatmap of differentially expressed genes | ww-deseq2 |
| `multiqc_report` | Aggregated interactive HTML QC report | ww-multiqc |
| `multiqc_data` | MultiQC parsed metrics data directory | ww-multiqc |

## Resource Considerations

### Compute Requirements
- **Memory**: 64GB recommended for human genome alignment (6GB minimum for chr1 testing)
- **CPUs**: 8+ cores recommended for efficient processing (2 cores minimum for testing)
- **Storage**: Sufficient space for reference genome, STAR index, BAM files, and analysis outputs
- **Network**: Stable internet connection for module imports and SRA downloads

### Optimization Tips
- Use call caching to save intermediate files if the pipeline crashes partway
- Adjust `star_cpu` and `star_memory_gb` parameters based on available resources
- For testing on limited hardware (e.g., GitHub Actions runners), use chr1 subset and reduce CPU/memory allocation
- Ensure adequate replicates per condition (3+ recommended) for robust DESeq2 analysis

> - Trim Galore defaults (quality=20, length=20) work well for most Illumina data

Reviewer: ```suggestion
```

I think these last three points aren't really about optimization and may not be helpful for users

## Testing the Pipeline

> The pipeline includes a test workflow that uses real RNA-seq data from the DESeq2 airway dataset:

Reviewer: ```suggestion
The pipeline includes a test workflow that uses real RNA-seq data referenced in the DESeq2 airway dataset:
```
I would clarify that the data didn't originate from the DESeq2 authors

5. Runs post-trim FastQC on trimmed reads
6. Builds STAR genome index
7. Performs two-pass alignment for all samples
8. Converts GTF to BED12 format for RSeQC
9. Runs RSeQC quality control on aligned BAMs
10. Combines gene count matrices from all samples
11. Performs DESeq2 differential expression analysis
12. Aggregates all QC reports with MultiQC

**Test Dataset Details:**
- Uses 4 samples from SRA (SRR1039508, SRR1039509, SRR1039512, SRR1039513)
- Real paired-end RNA-seq from human airway smooth muscle cells
- 2 untreated samples + 2 dexamethasone-treated samples
- Aligns to chromosome 1 subset for efficiency
- Resource requirements: 2 CPUs / 6GB RAM (fits GitHub Actions runners)

## Integration Patterns

This pipeline demonstrates several key WDL patterns:
- **Module Composition**: Combining 7 analysis modules in a coherent workflow
- **Data Passing**: Seamless transfer of outputs between QC, trimming, alignment, and analysis
- **Struct Usage**: Efficient organization of complex sample and reference data
- **Resource Management**: Coordinated resource allocation across modules
- **Scatter-Gather**: Parallel processing of multiple samples with aggregation

> - **QC Aggregation**: Collecting reports from multiple tools into a single MultiQC summary

Reviewer: ```suggestion
```
I don't think this section is needed since this isn't an educational pipeline

## Extending the Pipeline

This pipeline can be extended by:
- Adding UMI handling for deduplication
- Including pseudoalignment quantification (e.g., Salmon, Kallisto)
- Adding gene set enrichment analysis (GSEA)
- Including contamination screening
- Adding batch effect correction steps
- Integrating isoform-level analysis

## Related WILDS Components

- **ww-fastqc module**: Read quality assessment
- **ww-trimgalore module**: Adapter and quality trimming
- **ww-star module**: STAR alignment functionality
- **ww-bedparse module**: Genomic file format conversion
- **ww-rseqc module**: RNA-seq quality control
- **ww-deseq2 module**: Differential expression analysis
- **ww-multiqc module**: Aggregated QC reporting
- **ww-star-deseq2 pipeline**: Simpler RNA-seq pipeline (alignment + DE only)

> - **ww-sra-star pipeline**: SRA download to alignment pipeline

Reviewer: ```suggestion

```
The audience for these sections would be wdl developers I presume, but I don't think they're very helpful for users of the pipeline.
